### ISSUE TITLE
Escaping issue when using echo

### ISSUE THREAD

**Issue by Toflar** [andres-montanez/Magallanes#350]
Hey Andres,

I know this is rather a question but I just can't figure out the issue and I'm so confused now that I thought i'd just submit an issue and maybe you know why this isn't working immediately :-D

I have a task that writes updated content to my Symfony `parameters.yml` during the on-deploy stage. In Mage v1 this worked perfectly fine like this:

```php
$this->runCommandRemote("echo " . escapeshellarg($content) . " > app/config/parameters.yml");
```

I tried the very same for Nostromo but it just doesn't want to work in whatever way I try to get it to work :-(

```php
$process = $this->runtime->runRemoteCommand(
    sprintf('echo %s > app/config/parameters.yml', escapeshellarg($content)),
    true
);
```

The error message in the log file is the following:

```
bash: -c: line 0: unexpected EOF while looking for matching `"' bash: -c: line 8: syntax error: unexpected end of file
```

I noticed that you wrapped the whole `sh -c` command in `"` in v1 (https://github.com/andres-montanez/Magallanes/blob/1.0/Mage/Task/AbstractTask.php#L209) but you don't do this anymore in v3 anymore: https://github.com/andres-montanez/Magallanes/blob/master/src/Runtime/Runtime.php#L428

Is this due to something the Symfony Process component does internally?
Or do you see why this is not working?
Thank you for your help, really appreciate it :-)

**Comment by andres-montanez** [comment]
Hi, can you provide the command that is finally being executed?

**Comment by qzminski** [comment]
I have also asked about this on StackOverflow. Maybe any hint there will help to solve this issue.

https://stackoverflow.com/questions/42087135/sh-fails-to-write-content-to-a-file

**Comment by Toflar** [comment]
> Hi, can you provide the command that is finally being executed?

Wow, ultra fast :-) Sure, here's the log file (sensible data masked):

```
[2017-02-07 16:58:46] magephp.INFO: ssh -p 22 -q -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no user@domain.com sh -c \"cd /home/user/public_html/releases/20170207155839 \&\& echo 'parameters:     database_host: localhost     database_port: 3306     database_user: user     database_password: '\''***'\''     database_name: user     secret: ***     platform_version: develop ' > app/config/parameters.yml\" [] []
```

**Comment by andres-montanez** [comment]
Thanks, I will take a look ;)

**Comment by qzminski** [comment]
Just as an extra-info I can say it worked in the version 1 of mage but the command was executed in a different way there:

```
ssh foo@bar.com "sh -c \"cd dir && echo 'something' > file.txt\""
```

whereas the new version looks like this (note the missing wrapping double quotes):

```
ssh foo@bar.com sh -c \"cd dir \&\& echo 'something' > file.txt\"
```

I have tried to use the command syntax from the first version but then no output was caught by the Symfony Console component (when e.g. ran the ```cat app/config/parameters.yml``` command).

**Comment by andres-montanez** [comment]
Hi @Toflar can you test this patch?

```diff --git a/src/Runtime/Runtime.php b/src/Runtime/Runtime.php
index e03cc48..4b869fe 100644
--- a/src/Runtime/Runtime.php
+++ b/src/Runtime/Runtime.php
@@ -424,8 +424,8 @@ class Runtime
             $cmdDelegate = sprintf('cd %s && %s', $hostPath, $cmdDelegate);
         }
 
-        $cmdRemote = str_replace(['"', '&', ';', '|'], ['\"', '\&', '\;', '\|'], $cmdDelegate);
-        $cmdLocal = sprintf('ssh -p %d %s %s@%s sh -c \"%s\"', $sshConfig['port'], $sshConfig['flags'], $user, $host, $cmdRemote);
+        $cmdRemote = str_replace(['"', ';', '|'], ['\"', '\;', '\|'], $cmdDelegate);
+        $cmdLocal = sprintf('ssh -p %d %s %s@%s "%s"', $sshConfig['port'], $sshConfig['flags'], $user, $host, $cmdRemote);
 
         return $this->runLocalCommand($cmdLocal, $timeout);
     }
```
Let me know if it solves the issue, I've done some testings and it should be ok.
If everything works fine I will update the tests and push it to master, so I can fix the bug right away.

Regards.

**Comment by qzminski** [comment]
I can confirm that this patch solved the problem with the quotes and the echo command. However we also use the remote command ```pskill lsphp || true``` which clears the OPCache. This fails now because the pipe characters are being escaped. Do you think we can prevent escaping them as well?

**Comment by Toflar** [comment]
I wonder why you are escaping anything else than `"`. I think it should be as simple as

```php
$cmdRemote = str_replace('"', '\"', $cmdDelegate);
```

**Comment by andres-montanez** [comment]
Thanks @qzminski ! Yes, originally && and || had to be escaped, I will remove || as well, it shouldn't be needed now.
@Toflar did this also solved your problem?
Thank you both!

**Comment by Toflar** [comment]
```php
$cmdRemote = str_replace('"', '\"', $cmdDelegate);
$cmdLocal = sprintf('ssh -p %d %s %s@%s "%s"', $sshConfig['port'], $sshConfig['flags'], $user, $host, $cmdRemote);
```

fixes all issues ;-)

**Comment by andres-montanez** [comment]
Great!
I will push this over the weekend.
Thanks!

**Comment by andres-montanez** [comment]
Hi all, I've released version 3.0.1 with this bugfix.
Regards!
